asynch/proto/streams/buffered.py

```diff
@@ -111,31 +111,47 @@ def __init__(self, reader: StreamReader, buffer_max_size: int = constants.BUFFER
         self.current_buffer_size = 0
         self.position = 0
 
+    def _reset_buffer(self):
+        self.position = 0
+        self.buffer = bytearray()
+
     async def _read_into_buffer(self):
         packet = await self.reader.read(self.buffer_max_size)
         self.buffer.extend(packet)
         self.current_buffer_size = len(self.buffer)
 
+    async def _update_buffer_if_full(self):
+        if self.position == self.current_buffer_size:
+            self._reset_buffer()
+            await self._read_into_buffer()
+
     def _read_one(self):
+        if not self.buffer and not self.position:
+            return b""
         packet = self.buffer[self.position]
         self.position += 1
         return packet
 
     async def read_varint(self):
-        if self.position == self.current_buffer_size:
-            self._reset_buffer()
-            await self._read_into_buffer()
         packets = bytearray()
         while True:
+            await self._update_buffer_if_full()
             packet = self._read_one()
             packets.append(packet)
             if packet < 0x80:
                 break
         return leb128.u.decode(packets)
 
-    def _reset_buffer(self):
-        self.position = 0
-        self.buffer = bytearray()
+    async def read_bytes(self, length: int):
+        packets = bytearray()
+        while length > 0:
+            await self._update_buffer_if_full()
+            read_position = self.position + length
+            packet = self.buffer[self.position : read_position]  # noqa: E203
+            length -= len(packet)
+            self.position += len(packet)
+            packets.extend(packet)
+        return packets
```

Comment:
Just regrouping with addition of the `self._update_buffer_if_full` method. Is the name of this method is good enough or can be better - please share your variants.